lib/Analysis/Utils.h

```diff
@@ -0,0 +1,40 @@
+#ifndef LIB_ANALYSIS_UTILS_H_
+#define LIB_ANALYSIS_UTILS_H_
+
+#include "mlir/include/mlir/Analysis/DataFlowFramework.h"  // from @llvm-project
+#include "mlir/include/mlir/Interfaces/CallInterfaces.h"   // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"                // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+// A generalized version of visitExternalCall that joins all arguments to the
+// func.call op, and then propagates this joined value to all results. This is
+// useful for debug functions where there is a single operand that is not
+// changed by the external call, but can also be useful for some analyses like
+// secretness, where the result of the call is secret if any operand is secret.
+template <typename StateT, typename LatticeT>
+void visitExternalCall(CallOpInterface call,
+                       ArrayRef<const LatticeT *> argumentLattices,
+                       ArrayRef<LatticeT *> resultLattices,
+                       const std::function<void(AnalysisState *, ChangeResult)>
+                           &propagateIfChanged) {
+  StateT resultState = StateT();
+
+  for (const LatticeT *operand : argumentLattices) {
+    const StateT operandState = operand->getValue();
+    resultState = StateT::join(resultState, operandState);
+    if (resultState.isInitialized() && resultState.get()) {
+      break;
+    }
+  }
```

Comment:
I am OK with this now, the implicit conversion of `resultState.get()` to `bool` is a little bit unclear though.

In the future we might let the analysis define a default transfer function and reuse it in both `visitOperation` and `visitExternalCall`. For now we do not have the specific need for it.

Reply:
It's a good idea.

For this change, I think actually I will remove the break statement. Just blindly join all operands. This was originally a short-circuit optimization for secretness, knowing that once secretness is `true` every join will continue to produce `true`. For level/dimension this is not necessarily true.

Reply:
Removed the break